### ISSUE TITLE
Tested with 2021.2?

### ISSUE THREAD

**Issue by amirebrahimi** [transformsai/UnityGrpcWeb#2]
I tried making a build of this with Unity 2021.2 to WebGL and although it builds successfully when I run a very simple test I get this exception:

```
NotSupportedException: C:/Program Files/Unity 2021.2.11f1/Editor/Data/il2cpp/libil2cpp/icalls/mscorlib/System.Runtime.InteropServices/RuntimeInformation.cpp(34) : Unsupported internal call for IL2CPP:RuntimeInformation::GetRuntimeArchitecture - "This icall is not supported by il2cpp on this architecture."
Rethrow as TypeInitializationException: The type initializer for 'System.Runtime.InteropServices.RuntimeInformation' threw an exception.

Rethrow as TypeInitializationException: The type initializer for 'Grpc.Net.Client.Web.Internal.OperatingSystem' threw an exception.
```

Have you seen this before?

**Comment by MHDante** [comment]
Ah, I think this is due to a recent change in Grpc.Net.Client.Web, specifically: https://github.com/grpc/grpc-dotnet/pull/1171

in particular, it has to do with this one line: 
https://github.com/grpc/grpc-dotnet/blob/9f72e70c6a47ae823bb3ddcc719f1c63bf721fc7/src/Grpc.Net.Client/Internal/OperatingSystem.cs#L36


This issue shouldn't happen if you use the version of gRPC client libraries referenced by the package. It's likely that you either preinstalled the libs or you updated them:

Please verify that you have the following versions installed:
![image](https://user-images.githubusercontent.com/4652122/154810946-ddc48eb5-f202-49a9-bc1a-1e100e7e4a23.png)

Once again, I apologize profusely for the lack of docs.

----- 

That said, this will be an issue if we want to keep up to date with the official grpc libraries. 

Long term solutions: 

Best: 
The icall could be implemented in Unity il2cpp (even a stub that doesn't error would be nice) (@joshpeterson)

Okay: the static initializer linked above could have a failure tolerant case, or be set by client code. Perhaps it can be lazily initialized? (@JamesNK): 
This would have to be considered by the grpc-dotnet team. This may be difficult as this is not a platform they support.

Worst:
This [OperatingSystem class is primarily used in GrpcWebHandler](https://github.com/grpc/grpc-dotnet/blob/7d081d7c2682c792971d851e9e88d11a8728a2aa/src/Grpc.Net.Client.Web/GrpcWebHandler.cs#L37) (afaik). Since this library creates the GrpcWebHandler, I could explicitly set the IOperatingSystem object on it. Since this is currently an internal field, I would have to reflection-hack it. Maybe I could convince the grpc-dotnet team to expose it. 


I will leave this issue open as a way to track upstream changes (or if i decide to hack it)

**Comment by JamesNK** [comment]
> Ah, I think this is due to a recent change in Grpc.Net.Client.Web, specifically: https://github.com/grpc/grpc-dotnet/pull/1171

It's not a recent change. The change was over a year ago. Why is it only a problem now?

This is really a bug in Unity, but a work around that could be added to Grpc.Net.Client.Web is:
```cs
    internal class OperatingSystem : IOperatingSystem
    {
        public static readonly OperatingSystem Instance = new OperatingSystem();

        public bool IsBrowser { get; }

        private OperatingSystem()
        {
            try
            {
                IsBrowser = RuntimeInformation.IsOSPlatform(OSPlatform.Create("browser"));
            }
            catch
            {
                // IsOSPlatform isn't supported on Unity. Ignore exception.
                // https://github.com/transformsai/UnityGrpcWeb/issues/2
                IsBrowser = false;
            }
        }
    }
```

**Comment by MHDante** [comment]
Hey! You're right! I must admit that despite it being february, I thought it was 2021 still... I saw the PR and thought It was a recent change since the last library update. Big oof on my part.

After some more testing I found that the RuntimeInformation class initializer was indeed broken in unity 2021.
I'll submit a bug with unity and update this post when I can find out more. As for now, I suppose this library doesn't support unity 2021.2. Sorry for the misdirection, @amirebrahimi 

**Update:** The bug has been reported here:
https://fogbugz.unity3d.com/default.asp?1404506_qr4svffq1smd48cp

-----

In the meantime, I would really appreciate a trycatch on the operating system initializer.

As a note, this bug occurs mostly on the WebGL version of Unity. In this case, the IsBrowser field should report `true`, allowing the client to not override the browser's user-agent field. 

I wouldn't mind writing some logic in the GrpcWebHandler creation to set the flag manually.

**Comment by joshpeterson** [comment]
@MHDante: Thanks for the bug report - we will investigate it on the Unity side.

**Comment by MHDante** [comment]
The issue has been updated here: https://issuetracker.unity3d.com/issues/notsupportedexception-thrown-when-calling-any-member-of-system-dot-runtime-dot-interopservices-dot-runtimeinformation-in-webgl

Feel free to follow that link for updates

**Comment by ghost** [comment]
> @MHDante: Thanks for the bug report - we will investigate it on the Unity side.

We are experiencing this `Unsupported internal call for IL2CPP:RuntimeInformation::GetRuntimeArchitecture - "This icall is not supported by il2cpp on this architecture."` also. From investigating it seems to be a recurring issue from back in 2017.

@joshpeterson I don't suppose you have any information as to a current timeframe for this issue to be investigated on Unity's side?

Cheers!

**Comment by joshpeterson** [comment]
> I don't suppose you have any information as to a current timeframe for this issue to be investigated on Unity's side?

Thanks for raising this. I had forgotten about this conversation. It looks like this issue got stuck in the process and has not progressed much. I should be able to get it unstuck.

**Comment by ghost** [comment]
> Thanks for raising this. I had forgotten about this conversation. It looks like this issue got stuck in the process and has not progressed much. I should be able to get it unstuck.

Thanks for the update @joshpeterson. As a workaround for our project we will park our C# SDK implementation and start porting over to the JS SDK until we hear any more!

**Comment by joshpeterson** [comment]
I've fixed this issue internally now. It may take a day for it to show up on the public issue tracker, and a few weeks for the fixes to get through the back port process into Unity releases, but it is on the way.

**Comment by MHDante** [comment]
This issue has been Fixed in 2021.3.9f1, 2022.1.12f1, 2022.2.0b4, 2023.1.0a5

as described here: https://issuetracker.unity3d.com/issues/notsupportedexception-thrown-when-calling-any-member-of-system-dot-runtime-dot-interopservices-dot-runtimeinformation-in-webgl
